garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,206 @@
+"""
+DQN from https://arxiv.org/pdf/1312.5602.pdf.
+
+Known as Deep Q-Network, it estimates the Q-value function by deep neural
+networks. It enables Q-Learning to be applied on high complexity environments.
+To deal with pixel environments, numbers of tricks are usually needed, e.g.
+skipping frames and stacking frames as single observation.
+"""
+import numpy as np
+import tensorflow as tf
+
+from garage.logger import tabular
+from garage.misc.overrides import overrides
+from garage.misc.tensor_utils import normalize_pixel_batch
+from garage.tf.algos.off_policy_rl_algorithm import OffPolicyRLAlgorithm
+from garage.tf.misc import tensor_utils
+
+
+class DQN(OffPolicyRLAlgorithm):
+    """
+    DQN.
+
+    Args:
+        env_spec (garage.envs.env_spec.EnvSpec): Environment specification.
+        max_path_length (int): Maximum path length. The episode will terminate
+            when length of trajectory reaches max_path_length.
+        qf_lr (float): Learning rate for Q-Function.
+        qf_optimizer (tf.Optimizer): Optimizer for Q-Function.
+        discount (float): Discount factor for rewards.
+        name (str): Name of the algorithm.
+        target_network_update_freq (int): Frequency of updating target network.
+        grad_norm_clipping (float): How much to clip gradient. If None, no
```

Comment:
is `0.0` also no clipping -- if not, why not just use 0 as no clipping?
what is the scale of this parameter? is it [0,1], [0, 100], [-1,1]? i don't know.

Reply:
I guess not doing `tf.clip_by_norm` in a for loop will save some time, if we don't do gradient clipping.
```python
                if self.grad_norm_clipping is not None:
                    gradients = optimizer.compute_gradients(
                        self._loss, var_list=self.qf.get_trainable_vars())
                    for i, (grad, var) in enumerate(gradients):
                        if grad is not None:
                            gradients[i] = (tf.clip_by_norm(
                                grad, self.grad_norm_clipping), var)
                    self._optimize_loss = optimizer.apply_gradients(gradients)
                else:
                    self._optimize_loss = optimizer.minimize(
                        self._loss, var_list=self.qf.get_trainable_vars())
```

Reply:
i see, good point.

Reply:
please still update the comment with the proper range.